File: src/test/scala/nu/tengstrand/tetrisanalyzer/piecemove/PieceMoveTest.scala
package nu.tengstrand.tetrisanalyzer.piecemove

import nu.tengstrand.tetrisanalyzer.BaseTest
import org.junit.Test
import nu.tengstrand.tetrisanalyzer.board.Board
import nu.tengstrand.tetrisanalyzer.move.Move
import nu.tengstrand.tetrisanalyzer.piece.{PieceI, PieceS}

class PieceMoveTest extends BaseTest {

  @Test def setPiece {
    val board = Board(8,4)
    val piece = new PieceS
    val move = Move(0,3, 1)

    PieceMove(board, piece, move).setPiece()

    board should be (Board(Array(
      "#--------#",
      "#----xx--#",
      "#---xx---#",
      "#--------#",
      "##########")))
  }

  @Test def setPiece_clearTwoLines {
    val board = Board(Array(
      "#----------#",
      "#----x-----#",
      "#xxxxxxxxxx#",
      "#-x--x----x#",
      "#xxxxxxxxxx#",
      "############"))
    val piece = new PieceI
    val move = Move(1,4, 1)

    PieceMove(board, piece, move).setPiece() should be (2)

    board should be (Board(Array(
      "#----------#",
      "#----------#",
      "#----------#",
      "#----x-----#",
      "#-x--x----x#",
      "############")))
  }

  @Test def clearPiece {
    val board = Board(Array(
      "#xxxxxxxx#",
      "#xxxxxxxx#",
      "#xxxxxxxx#",
      "#xxxxxxxx#",
      "##########"))
    val piece = new PieceS
    val move = Move(0,3, 1)

    PieceMove(board, piece, move).clearPiece

    board should be (Board(Array(
      "#xxxxxxxx#",
      "#xxxx--xx#",
      "xxxx--xxx#",
      "#xxxxxxxx#",
      "##########")))
  }

  @Test def isFree {
    val board = Board(Array(
      "#-xxxxxxx#",
      "#-xxxxxxx#",
      "#xxxx--xx#",
      "#xxx--xxx#",
      "##########"))

    val piece = new PieceS
    val move = Move(0,3, 2)

    PieceMove(board, piece, move).isFree should be (true)
  }

  @Test def isFree_occupied {
    val board = Board(Array(
      "#--------#",
      "#--------#",
      "#--------#",
      "#----x---#",
      "##########"))

    val piece = new PieceS
    val move = Move(0,3, 2)

    PieceMove(board, piece, move).isFree should be (false)
  }
}

File: langs/java/src/test/java/com/github/tetrisanalyzer/piecemove/PieceMoveTest.java
package com.github.tetrisanalyzer.piecemove;

import com.github.tetrisanalyzer.board.Board;
import com.github.tetrisanalyzer.move.Move;
import com.github.tetrisanalyzer.piece.Piece;
import com.github.tetrisanalyzer.settings.AtariGameSettings;
import com.github.tetrisanalyzer.settings.PieceSettings;
import org.junit.Test;

import static com.github.tetrisanalyzer.piece.Piece.createPieceI;
import static com.github.tetrisanalyzer.piece.Piece.createPieceS;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

public class PieceMoveTest {
    PieceSettings settings = new AtariGameSettings();

    @Test
    public void setPiece() {
        Board board = Board.create(8, 4);
        Piece piece = createPieceS(settings);
        Move move = new Move(0,3, 1);

        new PieceMove(piece, move).setPiece(board);

        assertEquals(Board.create(
                "|--------|",
                "|----xx--|",
                "|---xx---|",
                "|--------|",
                "¯¯¯¯¯¯¯¯¯¯"), board);
    }

    @Test
    public void setPiece_clearTwoRows() {
        Board board = Board.create(
                "|----------|",
                "|----x-----|",
                "|xxxxxxxxxx|",
                "|-x--x----x|",
                "|xxxxxxxxxx|",
                "¯¯¯¯¯¯¯¯¯¯¯¯");
        Piece piece = createPieceI(settings);
        Move move = new Move(1,4, 1);

        assertEquals(2, new PieceMove(piece, move).setPiece(board));

        assertEquals(Board.create(
                "|----------|",
                "|----------|",
                "|----------|",
                "|----x-----|",
                "|-x--x----x|",
                "¯¯¯¯¯¯¯¯¯¯¯¯"), board);
    }

    @Test
    public void clearPiece() {
        Board board = Board.create(
                "|xxxxxxxx|",
                "|xxxxxxxx|",
                "|xxxxxxxx|",
                "|xxxxxxxx|",
                "¯¯¯¯¯¯¯¯¯¯");
        Piece piece = createPieceS(settings);
        Move move = new Move(0,3, 1);

        new PieceMove(piece, move).clearPiece(board);

        assertEquals(Board.create(
                "|xxxxxxxx|",
                "|xxxx--xx|",
                "|xxx--xxx|",
                "|xxxxxxxx|",
                "¯¯¯¯¯¯¯¯¯¯"), board);
    }

    @Test
    public void isFree() {
        Board board = Board.create(
                "|-xxxxxxx|",
                "|-xxxxxxx|",
                "|xxxx--xx|",
                "|xxx--xxx|",
                "¯¯¯¯¯¯¯¯¯¯");
        Piece piece = createPieceS(settings);
        Move move = new Move(0,3, 2);

        assertTrue(new PieceMove(piece, move).isFree(board));
    }
}
